src/io_ops/mod.rs

```diff
@@ -0,0 +1,103 @@
+use libc::{self, c_void, size_t};
+use log::error;
+
+use std::os::unix::prelude::AsRawFd;
+use std::{
+    ops::Deref,
+    os::unix::io::RawFd,
+    sync::{atomic::AtomicBool, Arc},
+};
+
+use std::io;
+
+#[derive(Debug, Clone)]
+pub struct ArcSubChannel(pub(crate) Arc<SubChannel>);
+
+impl ArcSubChannel {
+    pub fn as_raw_fd(&self) -> &FileDescriptorRawHandle {
+        self.0.as_ref().as_raw_fd()
+    }
+}
+
+impl Deref for ArcSubChannel {
+    type Target = SubChannel;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.as_ref()
+    }
+}
+#[async_trait::async_trait]
+impl crate::reply::ReplySender for ArcSubChannel {
+    async fn send(&self, data: &[&[u8]]) {
+        if let Err(err) = SubChannel::send(self.0.as_ref(), data).await {
+            error!("Failed to send FUSE reply: {}", err);
+        }
+    }
+}
+
+/// In the latest version of rust this isn't required since RawFd implements AsRawFD
+/// but until pretty recently that didn't work. So including this wrapper is cheap and allows
+/// us better compatibility.
+#[derive(Debug)]
+pub struct FileDescriptorRawHandle {
+    pub(in crate) fd: RawFd,
+    is_closed: AtomicBool,
+}
+
+impl FileDescriptorRawHandle {
+    pub fn new(fd: RawFd) -> Self {
```

Comment:
I think this function should be `unsafe`, and `fd` should private too, otherwise it can be used from safe code to close any arbitrary FD.